docs/source/specs/typespec/main.tsp

```diff
@@ -0,0 +1,251 @@
+        errors: ErrorMessage400[]
+    }
+
+    @error
+    model ErrorResponse400AlreadyExist<T>{
+        @statusCode
+        statusCode: 400,
+        errors: T[]
+    }
+
+    @error
+    model ErrorResponse403{
+        @statusCode 
+        statusCode: 403,
+        errors: ErrorMessage403[]
+    }
+
+    @error
+    model ErrorResponse404{
+        @statusCode 
+        statusCode: 404,
+        errors: ErrorMessage404[]
+    }
+
+    @error
+    model ErrorResponse401{
+        @statusCode 
+        statusCode: 401,
+    }
+
+    scalar ErrorCode extends string;
+}
+
+interface CreateResource<BodyType> {
+    create<ModelType>(@body body: BodyType): CreatedResponseForModel<ModelType> | 
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorResponse400 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface ListResource<T> {
+    list(...ListQueryParams): T[] |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface ReadResource<T> {
+    read(
+        @doc("Unique identification")
+        @path uuid: UUID;
+    ): T |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorResponse404 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface UpdateResource<B, R, E> {
+    update(
+        @doc("Unique identification")
+        @path uuid: UUID;
+        @body body: B
+    ): OkResponseForModel<R> |
+    Errors.ErrorResponse400AlreadyExist<E> |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface DeleteResource {
+    delete(
+        @doc("Unique identification")
+        @path uuid: UUID;
+    ): NoContentResponseResourceDeleted |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+@route("/workspaces")
+@tag("Workspaces")
+@doc("Operations about workspaces")
+namespace Workspaces {
+    model WorkspaceForRead {
+        @doc("Workspace A")
+        name: string = "Workspace A";
+        @doc("Description of Workspace A")
+        description?: string = "Description of Workspace A";
+    }
+
+    model Workspace {
+        @key uuid: UUID;
+        ...WorkspaceForRead;
+        ...Timestamps;
+    }
+
+    @error
+    model ErrorMessage400WorkspaceAlreadyExist {
+        detail?: string = "Workspace already exist in tenant",
+        status?: string = "400",
+        source?: string = "workspace",
+    }
+
+    alias CreateResourceAlias = CreateResource<WorkspaceForRead>;
+    @doc("Create workspace in tenant")
+    @summary("Create workspace in tenant")
+    @post op create is CreateResourceAlias.create<Workspace>;
+
+    model WorkspaceListResponse is ListResponse<Workspace>;
+    alias ListResourceAlias = ListResource<WorkspaceListResponse>;
+    @doc("By default, responses are sorted in ascending order by workspace name")
+    @summary("List workspaces for tenant")
+    @get op list is ListResourceAlias.list;
```

Comment:
I'm not sure how you add additional typed query parameters to this (probably a template parameter to ListResource, but see comment below about templating), but one of the things we probably need will need here is a "parent" parameter.

On the other hand, as mentioned above, we may want to omit this endpoint entirely for now and come back after our spikes.

Reply:
list action removed 👍